src/main/java/com/github/steveice10/packetlib/tcp/TcpClientSession.java

```diff
@@ -318,4 +318,15 @@ private static void createTcpEventLoopGroup() {
 
         Runtime.getRuntime().addShutdownHook(new Thread(() -> EVENT_LOOP_GROUP.shutdownGracefully()));
```

Comment:
I think this was supposed to functionally do the same thing. Do you know why it didn't work?

Reply:
The shutdown hook is called when the process has decided to shut down. It will decide to do so when the last non-daemon thread ends. However, the thread pools will not end unless instructed, so this hook will be called when it is no longer needed.

Reply:
Then should we mark the threads a daemon? Can we?

Reply:
If I understand the discussion in the mailing lists for NEtty they switched to a daemon thread model in Netty rel 5.0. I am not sure what other implications there are of making them daemon threads. Possibly no negative effect at all. We would have to change the thread factory. Or switch to Netty 5, but I have no idea how big a change that is.

Another option is to move away from static variables / thread pools and instead have per-instance groups. Since we anyway don't have more than one user of TcpClientSession in Geyser no pooling gain would be lost as there simpliy is no pooling gain.

Reply:
Moving to Netty 5 is not an option because Minecraft still uses Netty 4 and we depend on that in Geyser.
We have one TcpClientSession per GeyserSession class.

Reply:
I expected as much re netty 5.
Can we have more than one GeyserServer instance per process? If so, I guess it can be useful to have a global EventLoopGroup, but I'm not sure how useful it will be vs having one EventLoopGroup per TcpCLientSesstion instance which is shut down when the TcpClientSession instance is closed down.
It may be useful to test making a custom thread factory that creates daemon threads instead.

Reply:
It seems like we either remove this line, or we figure out the easiest way to make the threads daemon.

Reply:
Is the shutdown hook ever needed? Do we need to shut down the EventLoopGroup as a part of the shutdown, or is it sufficient to just let them be killed as a part of the shutdown?

Reply:
I specifically added the hook to try to mitigate this exact bug you're PRing to fix. So it sounds like this line is completely unnecessary.

Reply:
_If_ we decide to go with daemon threads then this line _may_ be needed to cleanly shut down the EventLoopGroup. I don't know. 

If we shut the EventLoopGroup down explicitly then this line is not needed.

I have not updated the patch pending directions from you about which way you want to take this.

Reply:
I don't know what the side effects are, but it seems like setting the threads to daemon means we don't need to fret about tasks holding up the server, and we don't need "new API". I think our Bedrock Protocol already marks its threads as daemon.

Reply:
I can make a test version and see what happens...

Reply:
See alternative deamon thread based approach in #748.